File: c/array_list.c
#include <stdio.h>

#define MAX 100

typedef struct {
    int data[MAX];
    int size;
} ArrayList;

void init(ArrayList *list) {
    list->size = 0;
}

void insert(ArrayList *list, int value) {
    if (list->size < MAX) {
        list->data[list->size] = value;
        list->size++;
    }
}

void deleteVal(ArrayList *list, int value) {
    int i, j;
    for (i = 0; i < list->size; i++) {
        if (list->data[i] == value) {
            for (j = i; j < list->size - 1; j++) {
                list->data[j] = list->data[j + 1];
            }
            list->size--;
            return;
        }
    }
}

void update(ArrayList *list, int index, int value) {
    if (index >= 0 && index < list->size) {
        list->data[index] = value;
    }
}

void traverse(ArrayList *list) {
    for (int i = 0; i < list->size; i++) {
        printf("%d ", list->data[i]);
    }
    printf("\n");
}

int main() {
    ArrayList list;

    init(&list);

    insert(&list, 10);
    insert(&list, 20);
    insert(&list, 30);

    traverse(&list);

    update(&list, 1, 25);
    traverse(&list);

    deleteVal(&list, 25);
    traverse(&list);

    return 0;
}

File: pascal/array_list.pas
program ArrayListImpl;

const
    MAX = 100;

type
    ArrayList = record
        data: array[0..MAX-1] of Integer;
        size: Integer;
    end;

var
    list: ArrayList;

procedure Init(var l: ArrayList);
begin
    l.size := 0;
end;

procedure Insert(var l: ArrayList; value: Integer);
begin
    if l.size < MAX then
    begin
        l.data[l.size] := value;
        Inc(l.size);
    end;
end;

procedure DeleteVal(var l: ArrayList; value: Integer);
var
    i, j: Integer;
begin
    for i := 0 to l.size - 1 do
        if l.data[i] = value then
        begin
            for j := i to l.size - 2 do
                l.data[j] := l.data[j + 1];
            Dec(l.size);
            Exit;
        end;
end;

procedure Update(var l: ArrayList; index, value: Integer);
begin
    if (index >= 0) and (index < l.size) then
        l.data[index] := value;
end;

procedure Traverse(var l: ArrayList);
var
    i: Integer;
begin
    for i := 0 to l.size - 1 do
        Write(l.data[i], ' ');
    WriteLn;
end;

begin
    Init(list);

    Insert(list, 10);
    Insert(list, 20);
    Insert(list, 30);

    Traverse(list);

    Update(list, 1, 25);
    Traverse(list);

    DeleteVal(list, 25);
    Traverse(list);
end.
